Add Visualization Components

mufeili opened the pull request.

jermainewang commented on minpy/nn/solver.py:
```diff
@@ -4,35 +4,39 @@
     models. The Solver performs stochastic gradient descent using different
     update rules defined in optim.py.
-
+    
```
Fix these trailing whitespaces. You could use `python ./dmlc-core/scripts/lint.py minpy python <your code>` to check for codestyle problems.

mufeili replied:
Checked, but returned "All passed!". Do you want me to delete all these whitespaces?

jermainewang replied:
That's weird. Just have a try in your forked repo:
```
 python ./dmlc-core/scripts/lint.py minpy python minpy/visualize/summaryOps.py                                    ──(Wed,Nov30)─┘
No config file found, using default configuration

************* Module summaryOps
 minpy/visualize/summaryOps.py:14: warning (W0311, bad-indentation, ) Bad indentation. Found 6 spaces, expected 8
 minpy/visualize/summaryOps.py:16: warning (W0311, bad-indentation, ) Bad indentation. Found 6 spaces, expected 8
 minpy/visualize/summaryOps.py:39: warning (W0311, bad-indentation, ) Bad indentation. Found 6 spaces, expected 8
 minpy/visualize/summaryOps.py:40: warning (W0311, bad-indentation, ) Bad indentation. Found 6 spaces, expected 8
 minpy/visualize/summaryOps.py:43: convention (C0330, bad-continuation, ) Wrong continued indentation (add 1 space).
                    if counts[i]]
                    ^|
 minpy/visualize/summaryOps.py:56: warning (W0311, bad-indentation, ) Bad indentation. Found 2 spaces, expected 4
 minpy/visualize/summaryOps.py:57: warning (W0311, bad-indentation, ) Bad indentation. Found 2 spaces, expected 4
 minpy/visualize/summaryOps.py:58: warning (W0311, bad-indentation, ) Bad indentation. Found 2 spaces, expected 4
 minpy/visualize/summaryOps.py:59: warning (W0311, bad-indentation, ) Bad indentation. Found 2 spaces, expected 4
 minpy/visualize/summaryOps.py:60: warning (W0311, bad-indentation, ) Bad indentation. Found 4 spaces, expected 8
 minpy/visualize/summaryOps.py:61: warning (W0311, bad-indentation, ) Bad indentation. Found 4 spaces, expected 8
 minpy/visualize/summaryOps.py:62: warning (W0311, bad-indentation, ) Bad indentation. Found 4 spaces, expected 8
 minpy/visualize/summaryOps.py:63: warning (W0311, bad-indentation, ) Bad indentation. Found 4 spaces, expected 8
 minpy/visualize/summaryOps.py:64: warning (W0311, bad-indentation, ) Bad indentation. Found 4 spaces, expected 8
 minpy/visualize/summaryOps.py:65: warning (W0311, bad-indentation, ) Bad indentation. Found 4 spaces, expected 8
 minpy/visualize/summaryOps.py:66: warning (W0311, bad-indentation, ) Bad indentation. Found 4 spaces, expected 8
 minpy/visualize/summaryOps.py:1: convention (C0103, invalid-name, ) Invalid module name "summaryOps"
 minpy/visualize/summaryOps.py:1: convention (C0111, missing-docstring, ) Missing module docstring
 minpy/visualize/summaryOps.py:7: error (E0401, import-error, ) Unable to import 'minpy'
 minpy/visualize/summaryOps.py:10: error (E0401, import-error, ) Unable to import 'minpy.visualization.summary_pb2'
 minpy/visualize/summaryOps.py:12: convention (C0103, invalid-name, scalarSummary) Invalid function name "scalarSummary"
 minpy/visualize/summaryOps.py:12: convention (C0103, invalid-name, scalarSummary) Invalid argument name "TAG"
 minpy/visualize/summaryOps.py:12: convention (C0111, missing-docstring, scalarSummary) Missing function docstring
 minpy/visualize/summaryOps.py:13: convention (C0121, singleton-comparison, scalarSummary) Comparison to True should be just 'expr' or 'expr is True'
 minpy/visualize/summaryOps.py:15: convention (C0121, singleton-comparison, scalarSummary) Comparison to True should be just 'expr' or 'expr is True'
 minpy/visualize/summaryOps.py:19: convention (C0103, invalid-name, histogramSummary) Invalid function name "histogramSummary"
 minpy/visualize/summaryOps.py:19: convention (C0103, invalid-name, histogramSummary) Invalid argument name "TAG"
 minpy/visualize/summaryOps.py:19: convention (C0111, missing-docstring, histogramSummary) Missing function docstring
 minpy/visualize/summaryOps.py:23: convention (C0103, invalid-name, _MakeHistogramBuckets) Invalid function name "_MakeHistogramBuckets"
 minpy/visualize/summaryOps.py:24: convention (C0103, invalid-name, _MakeHistogramBuckets) Invalid variable name "v"
 minpy/visualize/summaryOps.py:30: convention (C0103, invalid-name, _MakeHistogramBuckets) Invalid variable name "v"
 minpy/visualize/summaryOps.py:34: convention (C0103, invalid-name, _MakeHistogram) Invalid function name "_MakeHistogram"
 minpy/visualize/summaryOps.py:38: convention (C0103, invalid-name, _MakeHistogram) Invalid variable name "v"
 minpy/visualize/summaryOps.py:42: error (E0602, undefined-variable, _MakeHistogram) Undefined variable 'xrange'
 minpy/visualize/summaryOps.py:55: convention (C0103, invalid-name, WriteHistogramSeries) Invalid function name "WriteHistogramSeries"
 minpy/visualize/summaryOps.py:55: convention (C0103, invalid-name, WriteHistogramSeries) Invalid argument name "n"
 minpy/visualize/summaryOps.py:58: error (E0602, undefined-variable, WriteHistogramSeries) Undefined variable '_start_time'
 minpy/visualize/summaryOps.py:60: error (E0602, undefined-variable, WriteHistogramSeries) Undefined variable 'xrange'
 minpy/visualize/summaryOps.py:62: error (E0602, undefined-variable, WriteHistogramSeries) Undefined variable 'tf'
 minpy/visualize/summaryOps.py:62: error (E0602, undefined-variable, WriteHistogramSeries) Undefined variable 'tf'
 minpy/visualize/summaryOps.py:63: error (E0602, undefined-variable, WriteHistogramSeries) Undefined variable 'tf'
 minpy/visualize/summaryOps.py:6: convention (C0411, wrong-import-order, ) standard import "import bisect" comes before "import numpy"
 minpy/visualize/summaryOps.py:8: convention (C0411, wrong-import-order, ) standard import "import random" comes before "import numpy"
```

You should try to fix them. You could leave the `solver.py` file right now until @sneakerkg finished the lint fix on this file.

mufeili replied:
>  ************* Module summaryOps
>  
> =====1/1 python files passed check=====
> All passed!

Quite weird. Would you mind waiting until tomorrow? I will come tomorrow afternoon and maybe meeting face to face will be more helpful?

mufeili replied:
@jermainewang Also I haven't merged the master branch and visualize branch of my forked repo while I am trying to merge the visualize branch of my forked repo with the visualize branch of the minpy repo. If you try the files in the master branch of my forked repo, you will meet many problems that have been fixed in the visualize branch.

jermainewang commented on minpy/nn/solver.py:
```diff
@@ -178,23 +193,35 @@ def loss_func(*params):
         self.loss_history.append(loss.asnumpy())
 
+        if self.visualize:
+            Grad_norm = 0
```
variable name should not be uppercase in python

jermainewang commented on minpy/nn/solver.py:
```diff
@@ -243,11 +270,16 @@ def train(self):
             self.epoch = epoch + 1
             for each_batch in self.train_dataiter:
-                self._step(each_batch)
+                self._step(each_batch, t + 1)
```
just `t` ?

mufeili replied:
Fixed.

jermainewang commented on minpy/nn/solver.py:
```diff
@@ -243,11 +270,16 @@ def train(self):
+                if self.visualize:
+                    # Add scalar summaries of training loss.
+                    loss_summary = summaryOps.scalarSummary('loss', self.loss_history[-1])
+                    self.train_writer.add_summary(loss_summary, t + 1)
```
just `t`?

mufeili replied:
Since things are indexed from 0, I thought if this should start from 1. I can change back to t if you want.

jermainewang commented on minpy/visualize/event_file_writer.py:
```diff
@@ -0,0 +1,230 @@
+        A `bytes` object.
+    Raises:
+        TypeError: If `bytes_or_text` is not a binary or unicode string.
+    """
```
Refactor these comments in our style.

mufeili replied:
Fixed.

jermainewang commented on minpy/visualize/event_file_writer.py:
```diff
@@ -0,0 +1,230 @@
+        self._event.file_version = "brain.Event:2"
+        self.WriteEvent(self._event)
+
+    def WriteEvent(self, event):
```
Change the name of these functions to more pythonic style (no uppercases, underscores between words).

mufeili replied:
Fixed.

jermainewang commented on minpy/visualize/event_file_writer.py:
```diff
@@ -0,0 +1,230 @@
+        # the file_version of an event instance. As asserted in event.proto
+        # and events_writer.h, current version is "brain.Event:2" and all
+        # versions start with "brain.Event:".
+        self._event.file_version = "brain.Event:2"
```
remove ?

mufeili replied:
Removed.

jermainewang commented on minpy/visualize/event_file_writer.py:
```diff
@@ -0,0 +1,230 @@
+
+        self._num_outstanding_events = 0
+
+        self._recordio_writer = tf.python_io.TFRecordWriter(self._file_prefix)
```
Is this the ONLY part that depends on TF?

jermainewang commented on minpy/visualize/summaryOps.py:
```diff
@@ -0,0 +1,66 @@
+from minpy.visualization.summary_pb2 import Summary, HistogramProto
+
+def scalarSummary(TAG, scalar):
+    if isinstance(scalar, minpy.array.Array) == True:
```
Remove `== True`.

mufeili replied:
Removed

jermainewang commented on minpy/visualize/writer.py:
```diff
@@ -0,0 +1,214 @@
+# import graph_pb2
+import minpy.visualization.summary_pb2 as summary_pb2
+# import meta_graph_pb2
+import minpy.visualization.event_pb2 as event_pb2
```
Are these import correct? Not `minpy.visualize.event_pb2`?

mufeili replied:
Fixed.

jermainewang commented on minpy/visualize/writer.py:
```diff
@@ -0,0 +1,214 @@
+import minpy.visualization.event_pb2 as event_pb2
+# import meta_graph
+# from tensorflow.python.framework import ops
+# from tensorflow.python.platform import tf_logging as logging
```
Remove.

mufeili replied:
Removed.

jermainewang commented on minpy/visualize/writer.py:
```diff
@@ -0,0 +1,214 @@
+    meta_graph_bytes = meta_graph_def.SerializeToString()
+    event = event_pb2.Event(meta_graph_def=meta_graph_bytes)
+    self._add_event(event, global_step)
+  '''
```
remove

jermainewang commented on minpy/visualize/writer.py:
```diff
@@ -0,0 +1,214 @@
+    self.event_writer.add_event(event)
+
+
+class LegacySummaryWriter(SummaryToEventTransformer):
```
What's the meaning of this class? Can we just have one class?

mufeili replied:
Yes, we can.

mufeili replied:
Fixed.

mufeili commented:
@jermainewang 
How to generate _pb2.py files in installing script? What should I do for that?

jermainewang commented:
@hotpxl we have some protos borrowed from TF here. Two questions:
- What's the recommended way of generating _pb2.py files in setup?
- How to include the license of these codes? Put them in third_party and specify the license? Could you check TF's license to see whether we could use this?

jermainewang commented:
@mufeili Have you pushed the changes?

jermainewang commented:
One suggestions btw, please delete all these files in one commit next time.

mufeili commented:
I see. Thanks. @jermainewang

zihaolucky commented:
@mufeili Great work! BTW, do you have plan to ship these to MXNet?

mufeili commented:
@zihaolucky I can try if you want. I've just done the part of scalar summaries for Minpy and still exploring other kind of summaries. Do you want me to do the scalar part first or wait to have finished all the kinds of summaries?

zihaolucky commented:
@mufeili feel free to discuss here https://github.com/dmlc/mxnet/issues/4003. 

In my opinion, we need to support more summaries, that's the reason we choose tensorboard. But other summaries like audio and graph are not that trivial.

mufeili commented:
@jermainewang 

- Fixed problems found by lint.py in minpy/minpy/visualize/writer.py & summary_ops.py & event_file_writer.py

- Fixed problems found by lint.py in minpy/minpy/nn/solver.py due to my code. 

- Updated tutorials in minpy/docs/tutorial/visualization_tutorial, be sure to update the following pages accordingly: [Visualization with TensorBoard](http://minpy.readthedocs.io/en/latest/tutorial/visualization_tutorial/minpy_visualization.html#Logistic-regression) and [Visualization Solvers with TensorBoard](http://minpy.readthedocs.io/en/latest/tutorial/visualization_tutorial/minpy_visualization_adv.html)

zihaolucky commented:
Hi @jermainewang @mufeili 

As for the issue of the dependency on TF, I noticed the another api for summary/event in tensorflow v0.12:

```python
import tensorflow as tf

logs_path = 'log/experiment_logs'
writer = tf.summary.FileWriter(logs_path)

for i in range(100):
    summary = tf.Summary()
    summary.value.add(tag='counter', simple_value=float(i))

    # write log
    writer.add_summary(summary)

writer.close()
```

However when I check the `FileWriter` and its `EventFileWriter` inside, which calls a `pywrap_tensorflow.EventsWriter` to write a serilized event object to file. Do we also reply on a same `pywrap_tensorflow` in minpy? And it should be replaced by our own implementation as you've mentioned before?

mufeili commented:
@zihaolucky 

Hi Zihao. I've rewritten EventsWriter in python code. Currently the only dependency on TensorFlow is TFRecordWriter, check this: [minpy/minpy/visualize/event_file_writer.py](https://github.com/mufeili/minpy/blob/visualize/minpy/visualize/event_file_writer.py)

zihaolucky commented:
@mufeili Yep, the `event_file_writer` requires `tf.python_io.TFRecordWriter`, which is under `pywrap_tensorflow `. Is that right?

mufeili commented:
@zihaolucky 

Right. Strictly speaking, [tf.python_io.TFRecordWriter](https://github.com/tensorflow/tensorflow/blob/4172ca2cd7ac34be67bda2600944d284e8907b95/tensorflow/python/lib/io/tf_record.py) generates self._writer with pywrap_tensorflow.PyRecordWriter_New(
          compat.as_bytes(path), compat.as_bytes(compression_type), status)

zihaolucky commented:
@mufeili Thank you😊 

So if we want to bring this to MXNet, we might first figure out how to strip off the tensorflow component and write it on our own if necessary.

As I don't have experience in writing a wrapper(seems `pywrap_tensorflow` is a python wrapper for tf), could you give me some help? @mufeili @jermainewang Where can I start from if I want to replace this writer in tf by myself?

jermainewang commented:
@zihaolucky I think you could avoid the wrapper by reimplementing the logic merely in python. The summary part should not be a very costly part so I think python io is fine here.

zihaolucky commented:
@jermainewang Thank you, I will try.

As for the issue of generating `_pb2.py`, is this might help? http://stackoverflow.com/questions/27843481/python-project-using-protocol-buffers-deployment-issues

jermainewang commented on minpy/nn/solver.py:
```diff
@@ -4,10 +4,14 @@
 import minpy.numpy as np
 
+from minpy.visualize.writer import SummaryWriter
+import minpy.visualize.summary_ops as summary_ops
```
Is this working if no TF is installed?

mufeili replied:
Line 14 will work, line 13 won't. Maybe just leave the remark that uncomment this line when you have installed TensorFlow and want to visualize things?

jermainewang commented on minpy/nn/solver.py:
```diff
@@ -178,14 +193,27 @@ def loss_func(*params):
+            if self.visualize:
+                norm = dw ** 2
+                while not isinstance(norm, minpy.array.Number):
+                    norm = sum(norm)
```
`np.sum` ? What is this sum? This while also seems to weird. Will this cause dead loop?

jermainewang replied:
I would like to separate the visualize logic here to a separate `GradientVisualizer` object or class. It is an object or an string that is passed to the solver during initialization (rather than a bool flag). It is None by default, but could have `"l2norm"` as one option. Therefore, you could have `np.sum(dw ** 2)` implemented in another file under the same directory as `solver.py`. This makes the module much cleaner and support customized visualizer in the future.

mufeili replied:
@jermainewang Only a separate GradientVisualizer object or class? No need for other things like loss or accuracy?

jermainewang commented on minpy/nn/solver.py:
```diff
@@ -178,14 +193,27 @@ def loss_func(*params):
+            grad_norm_summary = summary_ops.scalar_summary('squared L2-norm of the gradient',
+                                                           grad_norm)
+            self.train_writer.add_summary(grad_norm_summary, iteration)
+
```
This part of logic could be simply put in the visualizer object I mentioned above.

jermainewang commented:
Also, please change the PR merge branch back to visualize branch.

mufeili commented:
@jermainewang 

- Create minpy.visualize.visualizer.py in the way of mxnet/optimizer.py
- Update minpy.nn.solver.py accordingly
- It works in the cases where visualizer is a string of "NoVisualize" or "Tensorboard" and where visualizer is a Visualizer object.
- Checked with lint.py

Let me know if there are any other issues or I can update the tutorials accordingly.

lryta commented:
@mufeili Thank you very much for your contribution Murphy. It is rewarding for our community. I will slightly change your code style and ask your help if necessary.

mufeili commented:
@lryta No problem Larry. It is interesting to have some idea about what it is like to make contributions to some real world projects. I cannot have this done without the help of @jermainewang and I've learnt a lot from him. I will now move on to see if I can add other summaries as well.

mufeili commented:
@lryta By the way, the original tutorials are now outdated and you need to update it, particularly  about the visualizer part. I can help about that part if you need after you change the code style part.

lryta commented:
@mufeili I definitely need your help. I've just stabilized in the US and I will let you know when I complete the style fix.

zzhang-cn commented:
We are releasing soon, so please Murphy and Larry work on this.

@Murphy: please let me know when and how I can take a look at your tutorial.

-zz

On Wed, Jan 11, 2017 at 2:15 AM, Larry Tang <notifications@github.com>
wrote:

> @mufeili <https://github.com/mufeili> I definitely need your help. I've
> just stabilized in the US and I will let you know when I complete the style
> fix.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/dmlc/minpy/pull/87#issuecomment-271653291>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AIKVUYeqZP2_Z_kbHHXEFB0Tmk0P2zGsks5rQ8qygaJpZM4LAtcj>
> .
>



-- 

*Zheng Zhang <https://shanghai.nyu.edu/academics/faculty/zheng-zhang> 张峥*

Professor of Computer Science, NYU  Shanghai 上海纽约大学计算机系教授


Room 1118, 1555 Century Avenue

Pudong New District, Shanghai 200122, China

中国上海市浦东新区世纪大道1555号1118室, 200122


T Shanghai office 上海办公室 +86 021 2059 5687
T China mobile 中国手机 +86 156 1808 6522
E 电子邮件 zz@nyu.edu

<jw5@nyu.edu>

murphy commented:
Wrong Murphy I’m afraid.

> On Jan 10, 2017, at 21:23, Zheng Zhang <notifications@github.com> wrote:
> 
> We are releasing soon, so please Murphy and Larry work on this.
> 
> @Murphy: please let me know when and how I can take a look at your tutorial.
> 
> -zz
> 
> On Wed, Jan 11, 2017 at 2:15 AM, Larry Tang <notifications@github.com>
> wrote:
> 
> > @mufeili <https://github.com/mufeili> I definitely need your help. I've
> > just stabilized in the US and I will let you know when I complete the style
> > fix.
> >
> > —
> > You are receiving this because you are subscribed to this thread.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/dmlc/minpy/pull/87#issuecomment-271653291>, or mute
> > the thread
> > <https://github.com/notifications/unsubscribe-auth/AIKVUYeqZP2_Z_kbHHXEFB0Tmk0P2zGsks5rQ8qygaJpZM4LAtcj>
> > .
> >
> 
> 
> 
> -- 
> 
> *Zheng Zhang <https://shanghai.nyu.edu/academics/faculty/zheng-zhang> 张峥*
> 
> Professor of Computer Science, NYU Shanghai 上海纽约大学计算机系教授
> 
> 
> Room 1118, 1555 Century Avenue
> 
> Pudong New District, Shanghai 200122, China
> 
> 中国上海市浦东新区世纪大道1555号1118室, 200122
> 
> 
> T Shanghai office 上海办公室 +86 021 2059 5687
> T China mobile 中国手机 +86 156 1808 6522
> E 电子邮件 zz@nyu.edu
> 
> <jw5@nyu.edu>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/dmlc/minpy/pull/87#issuecomment-271767924>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AF4BYqnuxApHH6vYCQFCIanEFcnpDpOeks5rREtMgaJpZM4LAtcj>.
>

mufeili commented:
@zzhang-cn I will do it immediately once Larry complete the style fix.
